modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-renderer/src/main/resources/META-INF/resources/js/thunks/fieldChange.es.js

```diff
@@ -0,0 +1,143 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import {debounce} from 'frontend-js-web';
+
+import {EVENT_TYPES} from '../actions/types.es';
+import {evaluate} from '../util/evaluation.es';
+import {PagesVisitor} from '../util/visitors.es';
+
+let REVALIDATE_UPDATES = [];
+
+const getEditedPages = ({editingLanguageId, name, pages, value}) => {
+	const pageVisitor = new PagesVisitor(pages);
+
+	return pageVisitor.mapFields(
+		(field) => {
+			if (field.name === name) {
+				return {
+					...field,
+					localizedValue: {
+						...field.localizedValue,
+						[editingLanguageId]: value,
+					},
+					value,
+				};
+			}
+
+			return field;
+		},
+		false,
+		true
+	);
+};
+
+const debounceFn = debounce((fn) => fn(), 50);
+
+export default function fieldChange({
+	defaultLanguageId,
+	editingLanguageId,
+	pages,
+	portletNamespace,
+	properties,
+	rules,
+}) {
+	return (dispatch) =>
+
+		// We added an event delay to break long tasks into short
+		// durations that are done when `fieldChange` is called
+		// very often. That's because changing a fieldEdited is a
+		// very expensive operation that causes many renderings in
+		// the application.
```

Comment:
Debouncing sounds like a good mitigation, but ideally we'd be able to reduce the root causes too (ie. reduce renders to essential renders only; and optimize slow updates to make them less expensive).

![just-sayin](https://user-images.githubusercontent.com/7074/86610617-e3c7f000-bfad-11ea-9ba2-b79ccc014fa9.gif)

Reply:
haha yeah, I was sad to have to use `debounce` here. I would expect to no longer use `debounce` for these things, but the current structure of the upper layers causes so much rendering that they need a more careful look and a restructuring of how they handle the data. And that would require more time to work on that ticket 🥺

And as I had commented on Slack about this performance problem when firing a fieldChange, it recreates a `pages`, and this has been causing rendering in all components. The data structure doesn't help us much here and also how the data stays on the Data Engine, where this performance problem was most noticeable